datafusion/core/src/physical_plan/sorts/builder.rs

```diff
@@ -15,136 +15,230 @@
 // specific language governing permissions and limitations
 // under the License.
 
-use arrow::compute::interleave;
-use arrow::datatypes::SchemaRef;
-use arrow::record_batch::RecordBatch;
 use datafusion_common::Result;
-use datafusion_execution::memory_pool::MemoryReservation;
+use std::collections::VecDeque;
+use std::mem::take;
 
-#[derive(Debug, Copy, Clone, Default)]
-struct BatchCursor {
-    /// The index into BatchBuilder::batches
-    batch_idx: usize,
+use super::cursor::Cursor;
+use super::stream::{BatchId, BatchOffset};
+
+pub type SortOrder = (BatchId, usize, BatchOffset); // batch_id, row_idx (without offset)
```

Comment:
what is the last element? the offset within the batch? How is that different than row index? Or perhaps the middle type is the number of rows?

Reply:
The BatchOffset comes from the cursor slices, as [discussed here](https://github.com/apache/arrow-datafusion/blob/d52049675ee5c335285635aea984d51b7cca9f74/datafusion/core/src/physical_plan/sorts/builder.rs#L170).

I'll also add a comment on the SortOrder type. TY!

Reply:
I did a little refactoring to make things clearer (hopefully). [This explainer](https://github.com/apache/arrow-datafusion/blob/3786021a2e087ee1576ab7be0becc01c75fa95e9/datafusion/core/src/physical_plan/sorts/batch_cursor.rs#L12-L20) now covers the batch offset.

Reply:
Removed the mutex, and updated the explainer.